python/cudf_polars/cudf_polars/experimental/scheduler.py

```diff
@@ -0,0 +1,148 @@
+    graph
+        The full task graph.
+
+    Returns
+    -------
+    List of other keys needed to extract ``key``.
+    """
+    keys = []
+    tasks = [graph[key]]
+    while tasks:
+        work = []
+        for w in tasks:
+            if istask(w):
+                work.extend(w[1:])
+            else:
+                try:
+                    if w in graph:
+                        keys.append(w)
+                except TypeError:  # not hashable
+                    pass
+        tasks = work
+    return keys
+
+
+def toposort(graph: Graph) -> list[Key]:
+    """Return a list of task keys sorted in topological order."""
+    # Stack-based depth-first search traversal. This is based on Tarjan's
+    # method for topological sorting (see wikipedia for pseudocode)
+    ordered: list[Key] = []
+    completed: set[Key] = set()
+    seen: set[Key] = set()
+
+    dependencies: Mapping[Key, list[Key]] = {k: required_keys(k, graph) for k in graph}
+    for key in graph:
+        if key in completed:
+            continue
+        nodes: list[Key] = [key]
+        while nodes:
+            # Keep current node on the stack until all descendants are visited
+            current = nodes[-1]
+            if current in completed:  # pragma: no cover
+                # Already fully traversed descendants of current
+                nodes.pop()
+                continue
+            seen.add(current)
+
+            # Add direct descendants of current to nodes stack
+            next_nodes = []
+            for nxt in dependencies[current]:
+                if nxt not in completed:
+                    if nxt in seen:  # pragma: no cover
+                        # Cycle detected!
+                        raise RuntimeError(
+                            f"Cycle detected in the task graph! Key: {nxt}"
+                        )
+                    next_nodes.append(nxt)
+
+            if next_nodes:
+                nodes.extend(next_nodes)
+            else:
+                # Current has no more descendants to explore
+                ordered.append(current)
+                completed.add(current)
+                seen.remove(current)
+                nodes.pop()
+
+    return ordered
+
+
+def synchronous_scheduler(
+    graph: Graph, key: Key, cache: MutableMapping | None = None
+) -> Any:
+    """
+    Execute the task graph for a given key.
+
+    Parameters
+    ----------
+    graph
+        The task graph to execute.
+    key
+        The final output key to extract from the graph.
+    cache
+        Intermediate-data cache.
+
+    Returns
+    -------
+    Executed task-graph result for ``key``.
+    """
+    if key not in graph:  # pragma: no cover
+        raise KeyError(f"{key} is not a key in the graph")
+    if cache is None:
+        cache = {}
+    for k in toposort(graph):
+        task = graph[k]
+        result = _execute_task(task, cache)
+        cache[k] = result
+    return _execute_task(key, cache)
```

Comment:
This seems to hold on to all intermediate results until the final answer is returned. This seems wrong since the size of the intermediates could be arbitrarily large.

Reply:
Yeah, that's fair. I suppose this is copying the "old" single-threaded scheduler used by dask.

Reply:
Okay, thanks for catching this. I updated tis function to ref-count the dependencies and clear items from the cache as soon as passible.